doc/developer/design/20230531_system_privileges.md

```diff
@@ -0,0 +1,86 @@
+
+- Associated: [RBAC](20230216_role_based_access_control.md)
+
+## Context
+
+Role based access control comprises a set of features that allow customers to control the
+privileges of individual users. Object privileges control user's access to database objects, they
+are granted with the `GRANT` keyword and revoked with the `REVOKE` keyword. For example,
+`GRANT SELECT ON TABLE t TO r` will grant the ability to read table `t` to role `r`. `GRANT CREATE
+ON SCHEMA s TO r` will grant the ability to create new objects in schema `s` to role `r`. Object
+privileges can be inherited through role membership. Role attributes control the user's ability to
+create (and sometimes modify) top level objects, they are granted and revoked with the `ALTER ROLE`
+keywords. For example, `ALTER ROLE r CREATEDB` will grant the ability to create databases to
+role `r`. `ALTER ROLE r CREATECLUSTER` will grant the ability to create clusters to role `r`. Role
+attributes cannot be inherited through role membership.
+
+Role attributes present some problems as they currently exist:
+
+- It can be confusing that there are two separate privilege primitives that control a role's
+  privilege.
+- Role attributes are not inherited which makes it cumbersome to grant attributes to a large
+  number of roles.
+
+## Goals
+
+- Reduce confusion around privileges.
+- Allow the ability to create databases, clusters, and roles to be inherited through role
+  membership.
+- Maintain PostgreSQL compatibility.
+
+## Overview
+
+We will add a new type of privilege called System Privileges. They will look and act just like
+normal object privileges, but they will grant the abilities that would otherwise be granted through
+role attributes. They will exist in parallel with role attributes.
+
+NOTE: Other databases, such as CockroachDB and Snowflake, have some form of system privileges that
+control similar privileges to role attributes.
+
+## Detailed description
+
+System privileges will be granted with the following syntax:
+`GRANT <privilege-specification> ON SYSTEM TO <role-specification>`
+
+System privileges will be revoked with the following syntax:
+`REVOKE <privilege-specification> ON SYSTEM TO <role-specification>`
+
+System privileges and role attributes will interact in the following ways:
+
+- Granting a system privilege will grant the privilege AND the attribute.
+- Revoking a system privilege will revoke the privilege AND the attribute.
+- Altering a role attribute will affect the role attribute but will not change the system privilege.
+- When checking to see if a role is allowed to execute a command we will check if they have the
+  system privilege or the role attribute.
+
+This approach has the following nice properties:
+
+- The ability to use system privileges without thinking about role attributes.
+- The ability to use role attributes without thinking about system privileges.
+- System privileges always take priority over role attributes when using both.
```

Comment:
Can system privs and role attrs ever be in conflict, or is it always safe to union them?

Reply:
It depends what you mean by in conflict. There is a possible confusing state where a role attribute can be false AND a system privilege exists. In this case if you only look at the role attribute you may think that the role is not allowed to do something when in fact the system privilege does allow the role to do that thing, because the system privilege takes precedent.

Here's a made up example:
```
CREATE ROLE r1;

ALTER ROLE r1 NOCREATEDB;

GRANT CREATEDB ON SYSTEM TO r1;
```

The attribute and the system privilege may be in conflict, but because the system privilege takes precedent then we go with the system privilege. This is one of the confusing/messy parts of trying to keep both.

Reply:
> Do you have idea/example of a use case where role attributes would be needed (aka why the postgres compatibility matters)?

Is this a question for me or @tr0njavolta?

Reply:
Ah sorry that question was for you @jkosh44 . E.g. in our docs I could imagine we say "role attributes are only necessary if you need to be postgres compatible. For example if you were <insert reason here>." I'm imagining that's something like "porting over your rbac configuration from your postgres database"?

Reply:
> "porting over your rbac configuration from your postgres database"

That's a good example I think. Another example is any tool that knows how to communicate with PostgreSQL and is expected role attributes to exist.